### PR TITLE
Adding parsing support for omp loop,  target loop directives

### PR DESCRIPTION
**Pull request by anchuraj**
Change adds parsing support for omp loop, omp target loop, omp target parallel loop and omp target teams loop.

### CODE REVIEW

File: llvm/include/llvm/Frontend/OpenMP/OMP.td [hunk 1]
```diff
@@ -2232,6 +2232,35 @@ def OMP_teams_loop : Directive<"teams loop"> {
   ];
   let leafConstructs = [OMP_Teams, OMP_loop];
 }
+def OMP_target_loop : Directive<"target loop"> {
```
Comment (ergawy) [inline llvm/include/llvm/Frontend/OpenMP/OMP.td:2235]: AFAIK, `target loop` is a non-standard construct, right? Where did we get the allowed clauses from? What `target loop` map to in the standard?

cc @mjklemm

**Review by ergawy — COMMENTED** [review]
Thanks @anchuraj! Just a question/clarification about `target loop`.

**Comment by mjklemm** [comment]
Yes, `target loop` is non-conforming.  But some vendor implementations support it and it will be coming with OpenMP 6.0.  So, we should accept it.  If people like, we may want to emit a warning it a non-conforming feature is being used and is being mapped to `target teams loop`.

**Comment by ergawy** [comment]
> Yes, `target loop` is non-conforming. But some vendor implementations support it and it will be coming with OpenMP 6.0. So, we should accept it. If people like, we may want to emit a warning it a non-conforming feature is being used and is being mapped to `target teams loop`.

@anchuraj sorry for the late review, yesterday was a public holiday.

I was more concerned about the list of allowed clauses and allowed once clauses. There are small differences between the 2 lists for target teams loop and target loop (for example `if` is in `allowedClauses` for `target teams loop` while in `allowedOnceClauses` for `target loop`). Maybe we just copy and paste the list from `target teams loop` or share them somehow.

**Review by ergawy — APPROVED** [review]

File: llvm/include/llvm/Frontend/OpenMP/OMP.td (hunk 1 above)
Comment (anchur) [reply llvm/include/llvm/Frontend/OpenMP/OMP.td:2235]: That is true. I copied the clauses from target and loop constructs.

**Comment by anchur** [comment]
> Yes, `target loop` is non-conforming. But some vendor implementations support it and it will be coming with OpenMP 6.0. So, we should accept it. If people like, we may want to emit a warning it a non-conforming feature is being used and is being mapped to `target teams loop`.

@mjklemm makes sense. I will add that

**Comment by anchur** [comment]
> > Yes, `target loop` is non-conforming. But some vendor implementations support it and it will be coming with OpenMP 6.0. So, we should accept it. If people like, we may want to emit a warning it a non-conforming feature is being used and is being mapped to `target teams loop`.
> 
> @anchuraj sorry for the late review, yesterday was a public holiday.
> 
> I was more concerned about the list of allowed clauses and allowed once clauses. There are small differences between the 2 lists for target teams loop and target loop (for example `if` is in `allowedClauses` for `target teams loop` while in `allowedOnceClauses` for `target loop`). Maybe we just copy and paste the list from `target teams loop` or share them somehow.

This part is confusing for me as well. For `target` construct it is specified in allowed once clause. I think it makes more sense to include it in allowed once as semantics of multiple if conditions included in a construct may be ambiguous. The following is from standard. I followed the same structure as specified in `target` and `loop` directives
```
Restrictions
20 Restrictions to the if clause are as follows:
21  At most one if clause can be specified that applies to the semantics of any construct or
22 constituent construct of a directive-specification.
```

**Comment by anchur** [comment]
Thank you @mjklemm  and @ergawy  for reviewing my PR. I was on vacation and sorry for the late response. Please let me know if there are other concerns.

**Comment by skatrak** [comment]
> > I was more concerned about the list of allowed clauses and allowed once clauses. There are small differences between the 2 lists for target teams loop and target loop (for example `if` is in `allowedClauses` for `target teams loop` while in `allowedOnceClauses` for `target loop`). Maybe we just copy and paste the list from `target teams loop` or share them somehow.
> 
> This part is confusing for me as well. For `target` construct it is specified in allowed once clause. I think it makes more sense to include it in allowed once as semantics of multiple if conditions included in a construct may be ambiguous.

I agree with you. It seems like the IF clause should be allowed once in all 3 cases: `target`, `target loop` and `target teams loop`. If `target teams loop` didn't do this already it seems like it's a bug because, as you pointed out, we can't have multiple IF clauses for the same construct, and in that combined construct only one of the leaves (`target`) can accept the clause.

File: llvm/include/llvm/Frontend/OpenMP/OMP.td [hunk 2]
```diff
@@ -2232,6 +2232,35 @@ def OMP_teams_loop : Directive<"teams loop"> {
...
+    VersionedClause<OMPC_FirstPrivate>,
+    VersionedClause<OMPC_IsDevicePtr>,
+    VersionedClause<OMPC_HasDeviceAddr, 51>,
+    VersionedClause<OMPC_LastPrivate>,
+    VersionedClause<OMPC_Map>,
+    VersionedClause<OMPC_Private>,
+    VersionedClause<OMPC_Reduction>,
+    VersionedClause<OMPC_UsesAllocators, 50>,
+    VersionedClause<OMPC_OMPX_Attribute>,
+    VersionedClause<OMPC_InReduction, 50>,
```
[the remark this answers is not in the document]
Comment (anchur) [reply llvm/include/llvm/Frontend/OpenMP/OMP.td:2248]: Yes, `allocate, defaultmap, depend, device, firstprivate, has_device_addr, if, in_reduction, is_device_ptr, map, nowait, private, thread_limit, uses_allocators` are clauses of target

**Review by mjklemm — APPROVED** [review]
LGTM

**Pull request merged**